Setting contentType for HEAD requests

megoth opened the pull request:
res.sendStatus(200) didn't pass the correct Content-Type, so this fix should handle that.

I think this could be handled better: we should not need to set Content-Type manually like this, that should be handled by the external http-handlers we use. But since get.js is kinda invasive in order to handle RDF-files, I think this is the "best" solution (for now).

RubenVerborgh commented:
> we should not need to set Content-Type manually like this

Note in this context #643 and related ones.

megoth commented:
This is a fix for #454 btw (forgot to link it earlier)

RubenVerborgh commented on lib/handlers/get.js:
```diff
@@ -76,7 +76,9 @@ function handler (req, res, next) {
     if (!includeBody) {
       debug('HEAD only')
-      res.sendStatus(200)
+      contentType = mime.contentType(_path.extname(path))
```
Not complete, since the current server assumes that extensionless paths are Turtle.

I would suggest to use an instance of LegacyResourceMapper here. It's overkill for now, but will give you the correct answer, and can in the future easily be replaced by ResourceMapper (which has a different MIME type detection system). See #661 and #662.

You can see an example instantiation in `lib/handlers/allow.js`.

megoth commented:
I've added a new test to handle extensionLess URLs, and made use of LegacyResourceMapper.

megoth commented:
I've stumbled upon a problem with extensionless files and NamedNode in rdflib. I've spoken with @kjetilk and we've decided that I'll put this a bit on hold (and do more research later on, hopefully with the help of @RubenVerborgh or @timbl ) so that we can focus on #177.

RubenVerborgh commented on lib/handlers/get.js:
```diff
@@ -76,7 +77,15 @@ function handler (req, res, next) {
       debug('HEAD only')
-      res.sendStatus(200)
+      const rootUrl = utils.getBaseUri(req)
+      const mapper = new LegacyResourceMapper({
```
Just wondering if you can initialize this earlier and reuse; fine otherwise.

megoth replied:
I could probably do that, this is more or less a hack to get it handled properly

RubenVerborgh commented on lib/handlers/get.js:
```diff
@@ -76,7 +77,13 @@ function handler (req, res, next) {
+        rootPath: ldp.root,
+        includeHost: ldp.multiuser
+      })
+      contentType = mapper._getContentTypeByExtension(path)
```
This is a private method though (underscore); you probably want `mapper.mapFileToUrl({ path }).contentType`.

megoth replied:
Ok, I'll do that ^_^

megoth commented:
@timbl @RubenVerborgh it's a little difficult for me to explain the problem, but I'll try.

The reason that the test breaks now is that `test/integration/ldp.js` -> `listContainer` -> `should ldp:contains the same files in dir` lists `sampleContainerblank` for the file `blank` (no extensions) instead of `blank`. 

I might be misunderstanding how it's used, but I think I can track the problem to an object (in a triple) that is set as NamedNode, and that doesn't parse the `memberUri` correctly. This is code in question (line ~60 or so in `lib/ldp-container.js`):

```
resourceGraph.add(
      resourceGraph.sym(containerUri),
      ns.ldp('contains'),
      resourceGraph.sym(memberUri))
```

`memberUri` in the case of the blank file is `https://server.tld/resources/sampleContainer/blank`, and that is not parsed correctly. The statement above becomes (I think):

```
Statement {
  subject:
   NamedNode {
     termType: 'NamedNode',
     value: 'https://server.tld/sampleContainer' },
  predicate:
   NamedNode {
     termType: 'NamedNode',
     value: 'http://www.w3.org/ns/ldp#contains' },
  object:
   NamedNode {
     termType: 'NamedNode',
     value: 'https://server.tld/sampleContainerblank' },
  why:
   NamedNode {
     termType: 'NamedNode',
     value: 'https://server.tld/sampleContainer' } }
```

So we see that the value of object is not what we expect.

Do I maybe need to call another method resourceGraph.sym in the case of an extensionless file? (I'll try this route next.)

timbl commented:
It looks to me as though memberUri is wrong

timbl commented:
Look at the code which generates the listing of the directory?  "https://server.tld/resources/sampleContainer/blank"  is at some point constructed from the URL the folder "https://server.tld/resources/sampleContainer/" and the filename "blank" -- or it is constructed from the URI of the foldder less the slash "https://server.tld/resources/sampleContainer" and "/" and "blank".   
Sounds like confusion as to wether the URI of the folder includes the slash
... in general it should so the code which generated the URI of the folder is wrong, or may
be this is a complex case where the client has been redirected from folder without slash to folder with slash

timbl commented:
Yes containerUri is wrong it should have a slash

megoth commented:
Finally got my head around back to this, and thanks to @timbl I fixed it ^_^ (so annoying when I get so focused on the wrong parts that I don't see the actual problem =P)

megoth commented:
Just to check; any reason to why `ResourceMapper.mapFileToUrl` was asynchronous? None of the functions it called was async, so I removed it and modified the places that used it accordingly.

Apart from that everything should be good for mergin now, @kjetilk ^_^

RubenVerborgh commented:
> Just to check; any reason to why `ResourceMapper.mapFileToUrl` was asynchronous?

Forward compatibility.

megoth commented:
Ok, then I'll make it asynchronous again ^_^

megoth commented:
New commits in place, that doesn't remove the asynchronous part of `mapFileToUrl`.

Looks good @RubenVerborgh ?

RubenVerborgh commented:
Fine with me!
